### DISCUSSION TITLE
enchanceEndpoints -> transformResponse TS error

### DISCUSSION THREAD

**Discussion by bogdan0083** [reduxjs/redux-toolkit#2726] [Q&A] [answered]
Hello guys! I'm trying to enhance the generated api and add `transformResponse` function to have normalized response:

```ts
import type {EntityState} from '@reduxjs/toolkit';
import {createEntityAdapter} from '@reduxjs/toolkit';
import type {
  GetTodosApiResponse,
  Todo as GeneratedTodo,
  TodosGetApiResponse,
} from '@shared/api/generatedApi';
import {generatedApi} from '@shared/api/generatedApi';

const todosAdapter = createEntityAdapter<Todo>();
const initialState = todosAdapter.getInitialState();

export type Todo = GeneratedTodo;

export const todosApiSlice = generatedApi.enhanceEndpoints<string>({
  addTagTypes: ['Todos'],
  endpoints: {
    getTodos: {
      // This property shows TypeScript error:
      // TS2322: Type '(response: Todo[]) => EntityState ' is not assignable to type 'undefined'.
      transformResponse(response: Todo[]): EntityState<Todo> {
        return todosAdapter.setAll(initialState, response);
      },
  },
});
```

The code **works**. The response get transformed but for some reason the ts compiler is not happy. Is there a way to fix ts error?

Thanks!

**Comment by phryneas** [comment, marked as answer]
```js
export const todosApiSlice = generatedApi.enhanceEndpoints<string>({
```
you should definitely not be putting that `<string>` there.

Apart from that, unfortunately currently `enhanceEndpoints` cannot change the type of the returned value.  It will work in runtime code, but they types will not accept it.
I'll see if I can change that somehow soon.

**Reply by bogdan0083** [reply]
> you should definitely not be putting that <string> there.

Yeah, that was not necessary.

> I'll see if I can change that somehow soon.

Thank you! That would help to use transformed data types. Right now it's impossible to use `selectFromResult` as well:

```ts
  const {data, error, isLoading} = useGetTodosQuery({_page: page}, {
    selectFromResult: (result) => ({
      ...result,
      // doesn't know about 'entities' property, obviously :)
      data: result.data ? Object.values(result.data.entities) : emptyArr,
    }),
  });
```

But that will be fixed automatically when we will have the ability to change the type of the returned value I guess.

**Reply by phryneas** [reply]
You can watch this in #2730
